source/python/idetect/nlp_models/category/category.py

```diff
@@ -0,0 +1,158 @@
+        tokens = [t for t in tokens if len(t) > 2]
+        tokens = [t for t in tokens if t not in stop_words]
+        tokens = [t.lower() for t in tokens]
+        tokens = [t for t in tokens if not t.isdigit()]
+        return tokens
+    
+    def fit(self, X, *args):
+        return self
+    
+    def transform(self, X, *args):
+        X = X.map(lambda x: self.prepare_tokens(x, self.stop_words))
+        return X
+
+class TfidfTransformer(TransformerMixin):
+    def __init__(self, no_below=5, no_above=0.5, tfidf_model=None, dictionary=None):
+        self.dictionary = dictionary
+        self.tfidf_model = tfidf_model
+        self.no_below = no_below
+        self.no_above = no_above
+    
+    def make_dictionary(self, texts):
+        if not self.dictionary:
+            self.dictionary = gensim.corpora.Dictionary(texts)
+            if self.no_below or self.no_above:
+                self.dictionary.filter_extremes(no_below=self.no_below, no_above=self.no_above)
+        return self
+    
+    def tfidf_transform(texts, dictionary=None, tfidf_model=None):
+        if not dictionary:
+            dictionary = gensim.corpora.Dictionary(texts)
+        corpus = [dictionary.doc2bow(text) for text in texts]
+        if not tfidf_model:
+            tfidf_model = gensim.models.TfidfModel(corpus)
+        corpus_tfidf = tfidf_model[corpus]
+        return corpus_tfidf, dictionary, tfidf_model
+    
+    def make_corpus(self, texts):
+        corpus = [self.dictionary.doc2bow(text) for text in texts]
+        return corpus
+    
+    def make_tfidf_model(self, corpus):
+        if not self.tfidf_model:
+            self.tfidf_model = gensim.models.TfidfModel(corpus)
+        return self
+    
+    def fit(self, texts, y=None):
+        self.make_dictionary(texts)
+        self.corpus = self.make_corpus(texts)
+        self.make_tfidf_model(self.corpus)
+        
+    def transform(self, texts):
+        corpus = self.make_corpus(texts)
+        return self.tfidf_model[corpus]
+
+class LsiTransformer(TransformerMixin):
+    def __init__(self, n_dimensions=100, no_below=5, no_above=0.5, lsi_model=None):
+        self.lsi_model = lsi_model
+        self.n_dimensions = n_dimensions
+        self.no_below = no_below
+        self.no_above = no_above
+    
+    def build_tfidf(self, texts):
+        self.tfidf_transformer = TfidfTransformer(no_below=self.no_below, no_above=self.no_above)
+        self.tfidf_transformer.fit(texts)
+        corpus_tfidf = self.tfidf_transformer.transform(texts)
+        dictionary = self.tfidf_transformer.dictionary
+        return corpus_tfidf, dictionary
+    
+    def lsi_to_vecs(self, corpus_lsi):
+        lsi_vecs = []
+        for c in corpus_lsi:
+            vec = [x[1] for x in c]
+            lsi_vecs.append(vec)
+        return np.array(lsi_vecs)
+    
+    def make_lsi_model(self, texts):    
+        self.corpus_tfidf, self.dictionary = self.build_tfidf(texts)
+        if not self.lsi_model:
```

Comment:
I actually removed this completely, it was just to accommodate a shortcut I was using when building models a while back.